[Map] Add Clustering Algorithms

smnandre opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Issues        | Fix #... 
| License       | MIT

Provide an interface and the first two PHP implementations of Clustering algorithms
- GridClustering
- Morton

Performances: < 1ms per 1000 Points *

* tested on a non production server (... my iMac 😅 )

**Next Step** will require JS code.

But this is already usefull with PHP classes only, for whomever has large amount of markers

smnandre committed:
[Map] Add Clustering Algorithms

smnandre commented:
(cannot ask for your review @Kocal as you're code owner, i would have otherwise ;) )

Kocal reviewed (commented):
Can you provide some examples/use-cases and documentation? I'm not really sure to understand what you are solving right now.

When you say "next step is JS", do you mean https://github.com/Leaflet/Leaflet.markercluster & https://developers.google.com/maps/documentation/javascript/marker-clustering?hl=fr implementation?

smnandre commented:
You have 30000 markers in database. Sending them all in HTML is non sense. Clustering them in backend is definitively the way to go. That way you only send data you want to display, and fix the performance problem some people had with too many infowindow for instance. 

Moreover it enforces the initial problem this bundle save: abstract the map services and manipulate data / UI from backend. 

People that do want to use custom GoogleMaps SDK Methods wont install this Bundle, so -personal opinion- it’s a mistake to base our choices on what they would do there.

Also this would not coexist with Live trait for now.

So Next step is to continue the path and provide the missing link: making the LiveComponent able to react on map events: zoom or pan, clicks, etc. 

That’s where I am right now, I Hope i’ve clarified my vision and goal for both this feature and more generally this bundle. 

And would love to integrate this Into yours, if you have a différent one :)

Kocal commented:
Thanks for the explanations! I was asking because I didn't see any modifications on `Map` nor in the documentation, and so I was wondering what you really wanted to do with this PR.

smnandre commented:
Yeah i'm sorry I miss time and need to push more often, or i end with hundreds of "features ready at 80%" on my computer and never find time to release them.

Here, it's because two unrelated people asked me for the same exact thing last week, I've reopened my "geo" archives and started to prepare the foundations. 

I'd like to end this (documentation included) before 2.24 😅

Kocal commented:
After a small rebase, that would be perfect 🤓

rvmourik commented:
Just a question out of curiosity, will this be in 2.24? Thanks in advance

mart-insiders commented:
Hi @rvmourik @Kocal , 
I was wondering what the current state of this feature is and what is required to finish it. 

At the moment some issues occur when implementing [https://github.com/Leaflet/Leaflet.markercluster](https://github.com/Leaflet/Leaflet.markercluster) with the current leaflet bridge.
The only possible way to add a leaflet layer with clustered markers at the moment is via javascript. 
This feels like unnecessary iterations over every marker in your stimulus controller.

`import { Controller } from "@hotwired/stimulus";

/* stimulusFetch: 'lazy' */
export default class extends Controller {
    connect() {
        this.element.addEventListener('ux:map:connect', this._onConnect);
    }

    disconnect() {
        this.element.removeEventListener('ux:map:connect', this._onConnect);
    }

    async _onConnect(event) {
      await import("leaflet.markercluster/dist/leaflet.markercluster-src.js");

      const { map } = event.detail;

      var clusterMarkerLayer = L.markerClusterGroup({
        spiderfyOnMaxZoom: true,
        showCoverageOnHover: false,
        zoomToBoundsOnClick: true
      });

      map.eachLayer(layer => {
        if (!(layer instanceof L.Marker)) {
            return;
        }

        clusterMarkerLayer.addLayer(layer);
        map.removeLayer(layer);
      })

      map.addLayer(clusterMarkerLayer);
    }
}`

An issue which occurs is Events are not persisted when creating a marker on this different layer.
Would the following changes suffice to resolve this?
- Implementation of a Layers object within the Map object or a reference in the Marker object. 
- Changing the `doCreateMarker` in `map_controller.js` to add a marker to a configured layer or as a layer to the map.

I would like to hear your opinion on this matter.

dannyvw commented:
@smnandre Do you perhaps have an update for this feature?

smnandre commented:
100% forgot this one : will rebase/clean tomorrow!

smnandre commented:
Cleaned a bit: 
- Cluster is now countable + iterable
- remove getCenterLat / getCenterLng with getCenter(): Point
- add PHP 8.1/2 support
- removed readonly classes
- rebased  and changelog updated

Not sure if doc will follow first, or if i first open the PR to integrate this in the components (depending on the time i find)

@Kocal : to me it's ready 

Thx @dannyvw for the poke :)

Kocal commented on src/Map/CHANGELOG.md:
```diff
@@ -3,6 +3,8 @@
 
 -  Add `minZoom` and `maxZoom` options to `Map` to set the minimum and maximum zoom levels
+-  Add `Cluster` class, interface for clustering algorithm and two implementations:
+   `GridClusteringAlgorithm` and `MortonClusteringAlgorithm`.
```
What about:
```suggestion
-  Add `Cluster` class and `ClusteringAlgorithmInterface` with two implementations `GridClusteringAlgorithm` and `MortonClusteringAlgorithm`.
```

?

Kocal commented on src/Map/src/Cluster/Cluster.php:
```diff
@@ -0,0 +1,84 @@
+        $this->points[] = $initialPoint;
+        $this->sumLat = $initialPoint->getLatitude();
+        $this->sumLng = $initialPoint->getLongitude();
+        $this->count = 1;
```
Maybe..? 
```suggestion
        $this->addPoint($initialPoint);
```

smnandre replied:
My code initially... but I had alert for division by zero risks in PhpStorm..

But in fact yeah that's better. Just need to add back initial values

Kocal commented on src/Map/src/Cluster/Cluster.php:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    /**
+     * @return Point[]
```
We can be more precise here: 
```suggestion
     * @return non-empty-list<Point>
```

Kocal commented on src/Map/tests/Cluster/ClusteringPerformanceTest.php:
```diff
@@ -0,0 +1,109 @@
+    public function testScenarioRegion50000(ClusteringAlgorithmInterface $algorithm, float $zoom): void
+    {
+        $points = $this->generatePoints(50000, 48.8, 49, 2.2, 2.5);
+        $this->runPerformanceTest($algorithm, $points, $zoom);
```
```suggestion
        $points = $this->generatePoints(50000, 48.8, 49, 2.2, 2.5);

        $this->runPerformanceTest($algorithm, $points, $zoom);
```

Kocal commented on src/Map/tests/Cluster/ClusteringPerformanceTest.php:
```diff
@@ -0,0 +1,109 @@
+    public function testScenarioWorld100000(ClusteringAlgorithmInterface $algorithm, float $zoom): void
+    {
+        $points = $this->generatePoints(100000, -90, 90, -180, 180);
+        $this->runPerformanceTest($algorithm, $points, $zoom);
```
```suggestion
        $points = $this->generatePoints(100000, -90, 90, -180, 180);

        $this->runPerformanceTest($algorithm, $points, $zoom);
```

Kocal commented on src/Map/tests/Cluster/GridClusteringAlgorithmTest.php:
```diff
@@ -0,0 +1,101 @@
+        // Expect two separate clusters
+        $clusters = $algorithm->cluster([$point1, $point2], 5.0);
+
+        $this->assertCount(2, $clusters);
```
I think it could have more value (feel free to modify the message if you want)
```suggestion
        $clusters = $algorithm->cluster([$point1, $point2], 5.0);

        $this->assertCount(2, $clusters, 'Two clusters should have created due to the high zoom value.');
```

Kocal replied:
(same comments for other similar code)

Kocal commented on src/Map/tests/Cluster/GridClusteringAlgorithmTest.php:
```diff
@@ -0,0 +1,101 @@
+        // Low zoom
+        $clusters = $algorithm->cluster([$point1, $point2], 1.0);
+
+        $this->assertCount(1, $clusters);
```
```suggestion
        $clusters = $algorithm->cluster([$point1, $point2], 1.0);

        $this->assertCount(1, $clusters, 'One cluster should have been created due to the low zoom value.');
```

smnandre commented on src/Map/src/Cluster/Cluster.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    private array $points = [];
+
+    private float $sumLat;
+    private float $sumLng;
+    private int $count;
```
```suggestion
    private float $sumLat = 0.0;
    private float $sumLng = 0.0;
    private int $count = 0;
```

smnandre commented:
Thanks for all the feedback / suggestions @Kocal : all commited :)

Kocal commented:
Rebased, conflicts fixed, and added some documentation.

Will merge when CI becomes green

Kocal commented:
Thank you @smnandre.
